src/rust/engine/async_semaphore/src/lib.rs

```diff
@@ -87,8 +107,12 @@ impl Drop for Permit {
     let task = {
       let mut inner = self.inner.lock();
       inner.available_permits += 1;
-      if let Some(task) = inner.waiters.pop_front() {
-        task
+      if let Some(unique_task) = inner.waiters.front() {
+        trace!(
```

Comment:
I think that this logging should maybe move outside of this crate and into a consumer. We have multiple semaphores, and logging in here won't allow us to identify them.

Reply:
It's not clear exactly where that would go. The speculating command runner makes sense because it knows the relative identities (primary, secondary) of the underlying runners, but then we wouldn't get any stats for non-speculation mode. I would like to put the logging in the specific command runners (local, remote), but they don't know anything about the bounded command running wrapping them.

Reply:
Then perhaps in the `BoundedCommandRunner`? Could give it a name for the runner. Not a blocker though.